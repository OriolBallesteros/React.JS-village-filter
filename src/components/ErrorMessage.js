import React, {Fragment} from 'react';

const ErrorMessage = props => {
    return (
        <Fragment>
            <p>Oops! Something went wrong. Try reloading again or show the next message to some developer:</p>
            <p>{props.error.data}</p>
            <p>Url fetched:</p>
            <p>{props.error.config.url}</p>
        </Fragment>
    );
};

export default ErrorMessage;


/**
 * PER RENDERITZAR NOMÃ‰S UN COMPONENT: MIRAR EXPENSLIST.JS DE ANDREW
 * 
 * 
 * <div>
        {
            props.expenses.length === 0 ? (
                <p>No expenses provided</p>
            ) : (
                    props.expenses.map((expense) => (
                        <ExpenseListItem {...expense} key={expense.id} />
                    ))
                )
        }
    </div>
 * 
 * 
 */