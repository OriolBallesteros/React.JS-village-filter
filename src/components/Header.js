import React from 'react';
import { NavLink } from 'react-router-dom';
import '../styles/Header.scss';

const Header = () => (
    <header className='Header'>
        <h1>Meet people in Brastlewark</h1>
        <NavLink to="/" className='NavLink' activeClassName="is-active" exact={true}>Go to filter</NavLink>
        <NavLink to="/searched" className='NavLink' activeClassName="is-active">Listed</NavLink>
    </header>
);

export default Header;