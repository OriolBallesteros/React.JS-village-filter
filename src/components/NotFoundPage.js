import React, { Fragment } from 'react';
import { Link } from 'react-router-dom';

const NotFoundPage = () => (
    <Fragment>
        404! - <Link to="/">Go home</Link>
    </Fragment>
);

export default NotFoundPage;