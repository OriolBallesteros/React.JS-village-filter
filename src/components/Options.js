import React from 'react';
import '../styles/Options.scss';

const Options = props => {
    return (
        <select className='Option' onChange={props.onHandleChange}>
            {props.list &&
                props.list.map(item => <option key={item} value={item}>{item}</option>)}
        </select>
    );
};

export default Options;