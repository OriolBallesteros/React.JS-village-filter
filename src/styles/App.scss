.App {
    text-align: center;
    padding: 0;

    h1{
        font-family: 'Srisakdi', cursive;
    }

    ul{
        list-style-type: none;
        padding: 0;
    }

    button {
        color: #fff !important;
        text-transform: uppercase;
        background: #00a3cc;
        padding: 20px;
        border-radius: 5px;
        display: inline-block;
        border: none;
    }

    button:hover {
        background: #003b4d;
        letter-spacing: 1px;
        box-shadow: 5px 40px -10px rgba(0,0,0,0.57);
        transition: all 0.4s ease 0s;
    
    }

    button:disabled{
        opacity: 0.5; 
        cursor: not-allowed;
    }
}
