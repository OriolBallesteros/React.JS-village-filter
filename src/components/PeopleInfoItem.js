import React, { Fragment } from 'react';
import { connect } from 'react-redux';
import Header from './Header';
import '../styles/PeopleInfoItem.scss';

const PeopleInfoItem = (props) => {
    const id = parseInt(props.match.params.id);
    const habitants = props.filteredHabitants;
    const habitant = habitants.filter((habitant) => {
        return habitant.id === id;
    })[0];

    return <PeopleInfoItemView habitant={habitant} />;
};

const PeopleInfoItemView = ({ habitant }) => (
    <Fragment>
        <Header className='PeopleInfoItem'/>
        <div className='PeopleInfoItem'>
            <h1>My name is {habitant.name}</h1>
            <img src={habitant.thumbnail} width="200" height="200" />
            <div>
                <p>I am {habitant.age} years old. My hair's color is: {habitant.hair_color}</p>
                <p>My height is: {habitant.height}, my weight: {habitant.weight}.</p>
            </div>

            {habitant.friends.length === 0 ?
                (<p>This person has no friends</p>) :
                (<div>
                    <p>My beloved friends are:</p>
                    <ul>
                        {habitant.friends.map((friend) => (
                            <li key={habitant.id + friend}>{friend}</li>
                        ))}
                    </ul>
                </div>)
            }

            <div>
                <p>I work as: </p>
                <ul>
                    {habitant.professions.map((prof) => (
                        <li key={habitant.id + prof}>{prof}</li>
                    ))}
                </ul>
            </div>
        </div>
    </Fragment>
);

const mapStateToProps = state => {
    return {
        filteredHabitants: state.filterReducer.filteredHabitants
    }
};

export const __PeopleInfoItemView = PeopleInfoItemView;
export default connect(mapStateToProps)(PeopleInfoItem);