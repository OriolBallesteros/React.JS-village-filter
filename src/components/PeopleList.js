import React, { Fragment } from "react";
import { connect } from "react-redux";
import PeopleListItem from "./PeopleListItem";
import Header from "./Header";
import "../styles/PeopleList.scss";

const PeopleList = props => {
  const { filteredHabitants } = props;

  return <PeopleListView filteredHabitants={filteredHabitants} />;
};

const PeopleListView = ({ filteredHabitants }) => (
  <Fragment>
    <Header />
    <div className="PeopleList">
      <p>Your search have {filteredHabitants.length} results</p>
      <p>Click the name to see more info about him or her</p>

      {filteredHabitants &&
        filteredHabitants.map(habitant => (
          <PeopleListItem key={habitant.id} habitant={habitant} />
        ))}
    </div>
  </Fragment>
);

const mapStateToProps = state => {
  return {
    filteredHabitants: state.filterReducer.filteredHabitants
  };
};

export const __PeopleListView = PeopleListView;
export default connect(mapStateToProps)(PeopleList);
