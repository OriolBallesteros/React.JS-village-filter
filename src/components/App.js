import React, { Component, Fragment } from "react";
import { connect } from "react-redux";
import { getVillage } from "../actions/apiActions";
import {
  setSelectedProfession,
  setSelectedHair,
  setSelectedGenre,
  setFilteredHabitants
} from "../actions/filterActions";

import Options from "./Options";
import ErrorMessage from "./ErrorMessage";
import "../styles/App.scss";

class App extends Component {
  componentWillMount() {
    this.props.getData();
  }

  //when selected option job
  onProfessionChange = e => {
    const profession = e.target.value;

    this.props.setSelectedProfession(profession);
  };

  //when selected option hair
  onHairChange = e => {
    const hair = e.target.value;

    this.props.setSelectedHair(hair);
  };

  //when selected option genre
  onGenreChange = e => {
    const genre = e.target.value;

    this.props.setSelectedGenre(genre);
  };

  //do filter
  handleSubmit = e => {
    e.preventDefault();

    this.props.setFilteredHabitants(
      this.props.apiInfo.village,
      this.props.filterInfo
    );

    //go to PeopleList component
    this.props.history.push("/searched");
  };

  render() {
    const { apiInfo, filterInfo } = this.props;
    let disableButton = true;
    if (filterInfo && filterInfo.profession && filterInfo.hair && filterInfo.genre) {
      disableButton = false;
    }

    return (
      <AppView
        apiInfo={apiInfo}
        filterInfo={filterInfo}
        disableButton={disableButton}
        that={this}
      />
      
    );
  }
}

const AppView = ({ apiInfo, filterInfo, disableButton, that }) => (
  <Fragment>
    {apiInfo.error && <ErrorMessage error={apiInfo.error} />}

    {!apiInfo.error && (
      <div className="App">
        <h1>Meet people in Brastlewark</h1>

        <form onSubmit={that.handleSubmit}>
          <div>
            <Options
              onHandleChange={that.onProfessionChange}
              list={apiInfo.professions}
            />

            <Options
              onHandleChange={that.onHairChange}
              list={apiInfo.hair_colors}
            />

            <Options
              onHandleChange={that.onGenreChange}
              list={["Male", "Female"]}
            />
          </div>
          <button disabled={disableButton}>Search</button>
        </form>

        {filterInfo && (
          <div>
            <p>(Remember you must select job, hair color and genre)</p>
            <ul>
              <li>
                Job:{" "}
                {filterInfo.profession
                  ? filterInfo.profession
                  : "You must select a profession"}
              </li>
              <li>
                Hair:{" "}
                {filterInfo.hair ? filterInfo.hair : "You must select a hair"}
              </li>
              <li>
                Genre:{" "}
                {filterInfo.genre
                  ? filterInfo.genre
                  : "You must select a genre"}
              </li>
            </ul>
          </div>
        )}
      </div>
    )}
  </Fragment>
);

const mapStateToProps = state => {
  return {
    apiInfo: state.apiReducer,
    filterInfo: state.filterReducer.selected
  };
};

const mapDispatchToProps = dispatch => {
  return {
    getData: () => { dispatch(getVillage()); },
    setSelectedProfession: prof => { dispatch(setSelectedProfession(prof)); },
    setSelectedHair: hair => { dispatch(setSelectedHair(hair)); },
    setSelectedGenre: genre => { dispatch(setSelectedGenre(genre)); },
    setFilteredHabitants: (village, selected) => { dispatch(setFilteredHabitants(village, selected)); }
  };
};

export const __AppView = AppView;
export default connect(mapStateToProps, mapDispatchToProps)(App);
