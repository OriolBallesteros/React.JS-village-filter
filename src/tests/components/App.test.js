import React from "react";
import { shallow } from "enzyme";
import { __AppView as AppView } from "../../components/App";

describe('App component tests', ()=>{
    it("should render App component when no error produced", () => {
        const apiInfo = {
            professions: ['prof1', 'prof2', 'prof3'],
            hair_colors: ['hair1', 'hair2', 'hair3']
        };
        const filterInfo = {
            profession: 'prof',
            hair: 'hair',
            genre: 'genre'
        };
        const disableButton = false;
        const that = {
            handleSubmit: jest.fn(),
            onProfessionChange: jest.fn(),
            onHairChange: jest.fn(),
            onGenreChange: jest.fn()
        };
    
        const wrapper = shallow(<AppView
            apiInfo={apiInfo}
            filterInfo={filterInfo}
            disableButton={disableButton}
            that={that} />);
    
        expect(wrapper).toMatchSnapshot();
    });
    
    it('should render ErrorMessage component when produced error on App component', ()=>{
        const apiInfo = {
            error: true
        };

        const wrapper = shallow(<AppView apiInfo={apiInfo}/>)

        expect(wrapper).toMatchSnapshot();
    });
});


