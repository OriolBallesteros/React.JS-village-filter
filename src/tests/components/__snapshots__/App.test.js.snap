// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component tests should render App component when no error produced 1`] = `
<Fragment>
  <div
    className="App"
  >
    <h1>
      Meet people in Brastlewark
    </h1>
    <form
      onSubmit={[MockFunction]}
    >
      <div>
        <Options
          list={
            Array [
              "prof1",
              "prof2",
              "prof3",
            ]
          }
          onHandleChange={[MockFunction]}
        />
        <Options
          list={
            Array [
              "hair1",
              "hair2",
              "hair3",
            ]
          }
          onHandleChange={[MockFunction]}
        />
        <Options
          list={
            Array [
              "Male",
              "Female",
            ]
          }
          onHandleChange={[MockFunction]}
        />
      </div>
      <button
        disabled={false}
      >
        Search
      </button>
    </form>
    <div>
      <p>
        (Remember you must select job, hair color and genre)
      </p>
      <ul>
        <li>
          Job:
           
          prof
        </li>
        <li>
          Hair:
           
          hair
        </li>
        <li>
          Genre:
           
          genre
        </li>
      </ul>
    </div>
  </div>
</Fragment>
`;

exports[`App component tests should render ErrorMessage component when produced error on App component 1`] = `
<Fragment>
  <ErrorMessage
    error={true}
  />
</Fragment>
`;
